File: leet/0127_word_ladder/0127_word_ladder.py
# LeetCode 127. Word Ladder
# https://leetcode.com/problems/word-ladder/
# O(M^2*N) time and space
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        self.words = set(wordList)
        if endWord not in self.words:
            return 0
        queue = deque([beginWord])
        steps = 2
        while queue:
            for _ in range(len(queue)):
                for neighbor in self.findNeighbors(queue.popleft()):
                    if neighbor == endWord:
                        return steps
                    queue.append(neighbor)
            steps += 1
        return 0
    
    def findNeighbors(self, word):
        chars = list(word)
        neighbors = []
        for i in range(len(word)):
            temp = chars[i]
            for c in string.ascii_lowercase:
                chars[i] = c
                newWord = "".join(chars)
                if newWord in self.words:
                    neighbors.append(newWord)
                    self.words.remove(newWord)
            chars[i] = temp
        return neighbors    

# O(M^2*N) time and space
class SolutionB:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        self.words = set(wordList)
        if endWord not in self.words:
            return 0
        self.words.remove(endWord)
        beginSet = set([beginWord])
        endSet = set([endWord])
        steps = 1
        while beginSet and endSet:
            if len(beginSet) > len(endSet):
                temp = endSet
                endSet = beginSet
                beginSet = temp
            newSet = set()
            for word in beginSet:
                if word in endSet:
                    return steps
                newSet.update(self.findNeighbors(word, endSet))
            beginSet = newSet
            steps += 1
        return 0
    
    def findNeighbors(self, word, endSet):
        chars = list(word)
        neighbors = []
        for i in range(len(word)):
            temp = chars[i]
            for c in string.ascii_lowercase:
                chars[i] = c
                newWord = "".join(chars)
                if newWord in self.words:
                    neighbors.append(newWord)
                    self.words.remove(newWord)
                if newWord in endSet:
                    neighbors.append(newWord)
            chars[i] = temp
        return neighbors    

File: leet/0127_word_ladder/0127_word_ladder.js
/*
LeetCode https://leetcode.com/problems/word-ladder/
O(w*l^2) time, w=words l=length of word, O(w*l) space
Key Insight: BFS graph problem, words different by 1 char are neighbors
*/
var ladderLength = function(beginWord, endWord, wordList) {
    const wordSet = new Set(wordList);
    if (!wordSet.has(endWord))
        return 0;
    const queue = [beginWord];
    let steps = 1, i = 0;
    while (i < queue.length) {
        for (sz = queue.length - i; sz > 0; sz--) {
            const word = queue[i++];
            if (word === endWord)
                return steps;
            findNeighbors(word, wordSet).forEach(nextWord => queue.push(nextWord));
        }
        steps++;
    }
    return 0;
};

function findNeighbors(word, wordSet) {
    const neighbors = [];
    for (let i = 0; i < word.length; i++) {
        for (const c of "abcdefghijklmnopqrstuvwxyz") {
            const nextWord = word.slice(0, i) + c + word.slice(i + 1);
            if (wordSet.has(nextWord)) {
                wordSet.delete(nextWord);
                neighbors.push(nextWord);
            }
        }
    }
    return neighbors;
}
